[Terraform]: Add resource_storage_transfer_job

akramhussein opened the pull request:
<!-- A summary of the changes in this commit goes here -->

<!--
Changes per downstream repository.  For each repository that you
expect to have changed, find the [tag] and write your commit
message beneath it.  More-specific tags replace less-specific tags.
For example, if you provide a message under [all], a message under
[puppet], and a message under [puppet-dns], the Terraform repository
will have the resulting commit made using the [all] message, the
Puppet Compute repository will have its commit made using [puppet],
and the Puppet DNS repository will have its commit made using
[puppet-dns].  You can delete unused tags, but you don't need to.

The structure of the PR body is important to our CI system!
The comments can be deleted, but if you want to make the downstream
commits sensible, you'll need to leave the dashed line separating
this PR's changes from the commit messages for downstream commits.
-->

<!-- 
Note: You may see "This branch is out-of-date with the base branch"
when you submit a pull request. This is fine! We don't use the GitHub
merge button to merge PRs, and you can safely ignore that message.
-->

-----------------------------------------------------------------
# [all]
## [terraform]
Add resource_storage_transfer_job

### [terraform-beta]
## [ansible]
## [inspec]

akramhussein commented:
Do I need to add the resource to `provider.go.erb` here?

rileykarson commented:
Yep! I haven't looked over this PR yet but you'll need to do that here as well.

akramhussein commented:
Done!

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,529 @@
+
+func resourceStorageTransferJob() *schema.Resource {
+	return &schema.Resource{
+
```
nit: remove this newline

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,529 @@
+			"description": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: func(v interface{}, k string) (ws []string, errors []error) {
```
[StringLenBetween](https://github.com/hashicorp/terraform/blob/master/helper/validation/validation.go#L151) can be used here instead I think? Ranging from [0, 1024]

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,529 @@
+	return []*schema.ResourceData{d}, nil
+}
+
+func objectConditions() *schema.Schema {
```
nit: Do you mind moving this + other subresource schemas right below the main schema?

rileykarson commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,185 @@
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccStorageTransferJob_basic(getTestProjectFromEnv()),
+				Check: resource.ComposeTestCheckFunc(
```
This resource supports import, so we can use "import-style tests" instead of a `Check`: See https://github.com/terraform-providers/terraform-provider-google/blob/2.0.0/google/resource_compute_address_test.go#L23 for an example.

akramhussein replied:
Does this mean it would simply become:

```
func TestAccStorageTransferJob_basic(t *testing.T) {
	t.Parallel()

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccStorageTransferJobDestroy,
		Steps: []resource.TestStep{
			{
				ResourceName:      "google_storage_transfer_job.foobar",
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
}
```

rileykarson replied:
Not quite, but almost!

```
func TestAccStorageTransferJob_basic(t *testing.T) {
	t.Parallel()

	resource.Test(t, resource.TestCase{
		PreCheck:     func() { testAccPreCheck(t) },
		Providers:    testAccProviders,
		CheckDestroy: testAccStorageTransferJobDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccStorageTransferJob_basic(getTestProjectFromEnv()),
			},
			{
				ResourceName:      "google_storage_transfer_job. transfer_job",
				ImportState:       true,
				ImportStateVerify: true,
			},
		},
	})
}
```

akramhussein replied:
Done

rileykarson commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,185 @@
+`, project, acctest.RandString(10), project, acctest.RandString(10), project, acctest.RandString(10), project)
+}
+
+func testAccStorageTransferJobExists(n string) resource.TestCheckFunc {
```
We don't need `Exists` checks if we're using import-style checks

akramhussein replied:
Removed `testAccStorageTransferJobExists`

rileykarson commented on third_party/terraform/utils/storage_transfer_helpers.go:
```diff
@@ -0,0 +1,312 @@
+	"time"
+)
+
+func expandDates(dates []interface{}) []*storagetransfer.Date {
```
nit: expand/flatten funcs should go in the main resource file below CRUD/import methods (at the very bottom generally)

akramhussein replied:
Moved to bottom and deleted the helpers file.

rileykarson commented on third_party/terraform/utils/storage_transfer_helpers.go:
```diff
@@ -0,0 +1,312 @@
+	return transferSpecsSchema
+}
+
+func validateDuration() schema.SchemaValidateFunc {
```
nit: this can live in [`validation.go`](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/utils/validation.go)

rileykarson commented on third_party/terraform/website/docs/r/storage_transfer_job.html.markdown:
```diff
@@ -0,0 +1,218 @@
+* `min_time_elapsed_since_last_modification` - (Optional)
+A duration in seconds with up to nine fractional digits, terminated by 's'. Example: "3.5s".
+
+* `include_prefixes` - (Optional If `include_refixes` is specified, objects that satisfy the object conditions must have names that start with one of the `include_prefixes` and that do not start with any of the `exclude_prefixes`. If `include_prefixes` is not specified, all objects except those that have names starting with one of the `exclude_prefixes` must satisfy the object conditions. See [Re ...
```
nit: missing `)` on optional

rileykarson commented on third_party/terraform/website/docs/r/storage_transfer_job.html.markdown:
```diff
@@ -0,0 +1,218 @@
+
+* `transfer_options` - (Optional) Characteristics of how to treat files from datasource and sink during job. If the option `delete_objects_unique_in_sink` is true, object conditions based on objects' `last_modification_time` are ignored and do not exclude objects in a data source or a data sink. Structure documented below.
+
+* `gcs_data_sink` - (Required) A Google Cloud Storage data sink. Structure documented below.
```
nit: `Required` params should be ordered before `Optional` ones

akramhussein commented:
Have also rebased on master and squashed this PR

akramhussein commented:
I signed it!

akramhussein commented:
Hi @rileykarson - I've rebased on master and also updated the authorship on this commit to match that for the signed CLA. Hopefully it should pick it up - it's done under `@project-sapiens.com`

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+			"description": {
+				Type:     schema.TypeString,
+				Required: true,
+				ValidateFunc: StringLenBetween(0, 1024),
```
```suggestion
				ValidateFunc: validation.StringLenBetween(0, 1024),
```

rileykarson commented on third_party/terraform/utils/validation.go:
```diff
@@ -187,6 +187,23 @@ func validateProjectName() schema.SchemaValidateFunc {
+			return
+		}
+
+		if _, err := time.ParseDuration(v); err != nil {
```
I think we need to import `time`

akramhussein commented:
Hi @rileykarson I'll work through these changes as individual commits and then rebase/squash.

Can I ask, what command are you using to compile?

rileykarson commented:
Feel free to keep separate commits/submit them however makes sense. The
problem before was the merge; our CI merge script rebases and then squashes
so it encountered a conflict before the merge that fixed it. I normally use
`git pull --rebase upstream master` to update old MM PRs (upstream is this
repo for me)

Assuming you have a checked out copy of Terraform in your GOPATH, MM can be
ran locally with:

bundle exec compiler -a -e terraform -o "${GOPATH}/src/
github.com/terraform-providers/terraform-provider-google/" -v "ga"

It will overwrite every file it generates in that directory! So make sure
you're on a clean checkout without uncommitted changes.

Beta can be generated by append -beta to the path & subbing beta for ga.
On Tue, Dec 18, 2018 at 10:01 AM Akram Hussein <notifications@github.com>
wrote:

> Hi @rileykarson <https://github.com/rileykarson> I'll work through these
> changes as individual commits and then rebase/squash.
>
> Can I ask, what command are you using to compile?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/magic-modules/pull/1069#issuecomment-448313117>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AB-_CMQL_e2HISPvklKxvKnVgf-g4uVEks5u6S2AgaJpZM4ZWrqe>
> .
>

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+						Required:  true,
+						Sensitive: true,
+					},
+					"secret_access_key": &schema.Schema{
```
Ah oops, meant to delete this one - deleting.

modular-magician commented:
I am a robot that works on MagicModules PRs!

I built this PR into one or more PRs on other repositories, and when those are closed, this PR will also be merged and closed.
depends: https://github.com/terraform-providers/terraform-provider-google-beta/pull/256 
depends: https://github.com/terraform-providers/terraform-provider-google/pull/2707

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+
+}
+
+func dateObject(required bool, optional bool) *schema.Schema {
```
To clarify, in the first suggestion is that keeping `dateObjectSchema` and having another function...i.e.:

```
func dateObjectSchema(required bool, optional bool) *schema.Schema {
	return &schema.Schema{
		Type:     schema.TypeList,
		Required: required,
		Optional: optional,
		ForceNew: true,
		MaxItems: 1,
		Elem: dateObjectResource(),
	}
}

func dateObjectResource() *schema.Resource {
	return &schema.Resource{
		Schema: map[string]*schema.Schema{
			"year": &schema.Schema{
				Type:         schema.TypeInt,
				Required:     true,
				ForceNew:     true,
				ValidateFunc: validation.IntBetween(0, 9999),
			},

			"month": &schema.Schema{
				Type:         schema.TypeInt,
				Required:     true,
				ForceNew:     true,
				ValidateFunc: validation.IntBetween(1, 12),
			},

			"day": &schema.Schema{
				Type:         schema.TypeInt,
				Required:     true,
				ForceNew:     true,
				ValidateFunc: validation.IntBetween(0, 31),
			},
		},
	},
}
```

If so, what would you call the second method?

rileykarson replied:
I would inline `dateObjectSchema` and give what you've called  `dateObjectResource` the name `dateObjectSchema`. Similar to https://github.com/terraform-providers/terraform-provider-google/blob/2.0.0/google/resource_dataproc_cluster.go#L391

akramhussein commented on third_party/terraform/website/docs/r/storage_transfer_job.html.markdown:
````diff
@@ -0,0 +1,218 @@
+$ terraform import google_storage_transfer_job.nightly-backup-transfer-job my-project-1asd32/8422144862922355674
+```
+
+Note that when importing a Transfer Job (and only when importing), the Compute API needs to be enabled - you'll see an error with a link to the enablement page if it is not.
````
It was quite a while back when I wrote original code and can't remember but definitely was an issue at the time. Not sure how I'd re-test without a new project. If I have time will look in to it.

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+	}
+
+	d.Set("name", res.Name)
+	jobId, err := extractTransferJobId(res.Name)
```
Is this ok - https://github.com/GoogleCloudPlatform/magic-modules/pull/1069/commits/e08d41b24cef17432aa7faa890167dc5c4c5dc6d

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+			ScheduleStartDate: expandDates([]interface{}{schedule["schedule_start_date"]})[0],
+		}
+
+		if v, ok := schedule["schedule_end_date"]; ok && len(v.([]interface{})) > 0 {
```
Like this?

```
func expandTransferSchedules(transferSchedules []interface{}) *storagetransfer.Schedule {
	if len(transferSchedules) == 0 || transferSchedules[0] == nil {
		return nil
	}

	schedule := transferSchedules[0].(map[string]interface{})
	return &storagetransfer.Schedule{
		ScheduleStartDate: expandDates([]interface{}{schedule["schedule_start_date"]})[0],
		ScheduleEndDate = expandDates([]interface{}{schedule["schedule_end_date"]})[0],
		StartTimeOfDay = expandTimeOfDays([]interface{}{schedule["start_time_of_day"]})[0]
	}
}
```

rileykarson replied:
That looks reasonable to me!

akramhussein replied:
I've now applied the same principle to the other expand and flatten functions.

akramhussein commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,141 @@
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccStorageTransferJob_basic(t *testing.T) {
```
Do you have a good example? `

rileykarson replied:
Like https://github.com/GoogleCloudPlatform/magic-modules/blob/master/third_party/terraform/tests/resource_access_context_manager_access_level_test.go.erb#L31, providing a second schema where a non-`ForceNew` attribute is modified.

akramhussein commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,141 @@
+	"github.com/hashicorp/terraform/terraform"
+)
+
+func TestAccStorageTransferJob_basic(t *testing.T) {
```
Other than running `make testacc TEST=./google`? I was going to setup a new project to test the `Compute API` issue below - did you have another way to run tests in mind?

rileykarson replied:
Nope, that's how we run them as well! I just wanted to confirm you were able to run them yourself in case of issues like quota.

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,808 @@
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
```
@rileykarson do you want me to apply this to all schema functions - to return `&schema.Resource`?

rileykarson replied:
Sure!

akramhussein commented:
Hi @rileykarson I believe I've now addressed all the comments except for the `update` unit test.
Would you be able to have a quick pass and let me know if everything looks good? Thanks for being patient! Nearly there (never written go before!)

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 9c6b135f) have been included in your existing downstream PRs.

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,728 @@
+func flattenAwsS3Data(awsS3Data *storagetransfer.AwsS3Data) []map[string]interface{} {
+	data := map[string]interface{}{
+		"bucket_name":    awsS3Data.BucketName,
+		"aws_access_key": awsS3Data.AwsAccessKey,
```
This has to call the flatten function I think?

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,728 @@
+func flattenAwsAccessKeys(awsAccessKey *storagetransfer.AwsAccessKey) []map[string]interface{} {
+	data := map[string]interface{}{
+		"access_key_id":     awsAccessKey.AccessKeyId,
+		"secret_access_key": awsAccessKey.SecretAccessKey,
```
You are correct, it's not returned https://cloud.google.com/storage-transfer/docs/create-manage-transfer-program

akramhussein commented:
Fixed the linting issue. Unfortunately during testing i hit a rate limit:

```bash
make testacc TEST=./google TESTARGS='-run=TestAccStorageTransferJob_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 TF_SCHEMA_PANIC_ON_ERROR=1 go test ./google -v -run=TestAccStorageTransferJob_basic -timeout 120m -ldflags="-X=github.com/terraform-providers/terraform-provider-google/version.ProviderVersion=acc"
=== RUN   TestAccStorageTransferJob_basic
=== PAUSE TestAccStorageTransferJob_basic
=== CONT  TestAccStorageTransferJob_basic
--- FAIL: TestAccStorageTransferJob_basic (239.03s)
    testing.go:518: Step 0 error: Error applying: 1 error(s) occurred:

        * google_storage_transfer_job.transfer_job: 1 error(s) occurred:

        * google_storage_transfer_job.transfer_job: Error reading Transfer Job "transferJobs/17748877939099650077": googleapi: Error 429: Quota exceeded for quota group 'ReadOnlyGroup' and limit 'USER-100s' of service 'storagetransfer.googleapis.com' for consumer 'project_number:278407580213'., rateLimitExceeded
FAIL
FAIL	github.com/terraform-providers/terraform-provider-google/google	239.079s
make: *** [testacc] Error 1
```

akramhussein commented:
Removed the line about requiring the `Compute API` - I did some testing and couldn't replicate it.

akramhussein commented:
Hi @rileykarson - I’ve added 2 more steps that 1) change data sink name; then 2) change the data source name. I’ve re-used the same template schema - didn’t see benefit of duplicating it.

I’ve tested these against our GCP account and all works. Think that's it?

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 22e2b942) have been included in your existing downstream PRs.

rileykarson commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,728 @@
+		fmt.Printf("Error creating transfer job %v: %v", transferJob, err)
+		return err
+	}
+
```
We should `SetId` in `Create` after a successful create; that way, if `Read` fails afterwards (such as to a `503`) Terraform will know we've created the resource and won't try recreating it.

rileykarson commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,165 @@
+  role          = "roles/storage.admin"
+  member        = "serviceAccount:${data.google_storage_transfer_project_service_account.default.email}"
+
+  depends_on    = [
```
This resource will already depend on those other resources because of the interpolations, so this block is redundant.

rileykarson commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,165 @@
+  role          = "roles/storage.admin"
+  member        = "serviceAccount:${data.google_storage_transfer_project_service_account.default.email}"
+
+  depends_on    = [
```
Same as above

rileykarson commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,165 @@
+		}
+	}
+
+	depends_on = [
```
For completeness, this is not redundant- Terraform doesn't know to wait for those IAM roles to be applied without this block.

akramhussein commented:
I've addressed all the points - can you just double check the `setId` calls?

The CLA was also signed and did go green but I think I pushed some intermediate commits before changing authorship and it got confused.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit a3097f86) have been included in your existing downstream PRs.

akramhussein commented on third_party/terraform/resources/resource_storage_transfer_job.go:
```diff
@@ -0,0 +1,730 @@
+	}
+
+	d.Set("name", res.Name)
+	d.SetId(fmt.Sprintf("%s/%s", project, res.Name))
```
Ok will do, but I had a method for this in my original PR called that you asked me to remove.

```
func extractTransferJobId(id string) (string, error) {
	if !regexp.MustCompile("^transferJobs/.+$").Match([]byte(id)) {
		return "", fmt.Errorf("Invalid transferJob id format, expecting transferJob/{id}")
	}
	parts := strings.Split(id, "/")
	return parts[1], nil
}
```

akramhussein replied:
No problemo - what is the command you use to test. I've been using `make testacc TEST=./google TESTARGS='-run=TestAccDataSourceGoogleStorageTransferProjectServiceAccount_basic`

rileykarson replied:
Afaik that command will test the datasource and not the resource? I've been using:

```
make testacc TEST=./google TESTARGS='-run=TestAccStorageTransferJob_basic'
```

akramhussein replied:
Argh sorry pasted wrong one!

I'm using:

```make testacc TEST=./google TESTARGS='-run=TestAccStorageTransferJob_basic'```

So that's fine.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit 729eaf7b) have been included in your existing downstream PRs.

akramhussein commented on third_party/terraform/tests/resource_storage_transfer_job_test.go:
```diff
@@ -0,0 +1,155 @@
+			},
+			{
+				Config: testAccStorageTransferJob_basic(getTestProjectFromEnv(), testDataSourceId, testUpdatedDataSinkId, testTransferJobId),
+				Check:  resource.TestCheckResourceAttr("google_storage_transfer_job.transfer_job.transfer_spec.0.gcs_data_sink.0", "bucket_name", testUpdatedDataSinkId),
```
should be `"google_storage_transfer_job.transfer_job", "transfer_spec.0.gcs_data_sink.0.bucket_name"` but can remove.

In terms of an update test for a non-`ForceNew`field - what should I write?

akramhussein commented:
Updated the tests and they pass now.

modular-magician commented:
I am (still) a robot that works on MagicModules PRs!

I just wanted to let you know that your changes (as of commit c70dccce) have been included in your existing downstream PRs.

akramhussein commented:
🎉 thanks!
